testsupport/tiers/checks.go

```diff
@@ -47,58 +46,123 @@ var (
-	switch tier {
+func newChecks(tierName string) (TierChecks, error) {
+	switch tierName {
 	case base:
 		return &baseTierChecks{tierName: base}, nil
-
 	case baselarge:
 		return &baselargeTierChecks{baseTierChecks{tierName: baselarge}}, nil
-
 	case baseextended:
 		return &baseextendedTierChecks{baseTierChecks{tierName: baseextended}}, nil
-
 	case baseextendedidling:
 		return &baseextendedidlingTierChecks{baseTierChecks{tierName: baseextendedidling}}, nil
-
 	case basedeactivationdisabled:
 		return &basedeactivationdisabledTierChecks{baseTierChecks{tierName: basedeactivationdisabled}}, nil
-
 	case hackathon:
 		return &hackathonTierChecks{baseTierChecks{tierName: hackathon}}, nil
-
 	case advanced:
 		return &advancedTierChecks{baseTierChecks{tierName: advanced}}, nil
-
 	case appstudio:
 		return &appstudioTierChecks{tierName: appstudio}, nil
-
 	case test:
 		return &testTierChecks{tierName: test}, nil
-
 	default:
-		return nil, fmt.Errorf("no assertion implementation found for %s", tier)
+		return nil, fmt.Errorf("no assertion implementation found for %s", tierName)
 	}
 }
 
 type TierChecks interface {
-	GetNamespaceObjectChecks(nsType string) []namespaceObjectsCheck
 	GetClusterObjectChecks() []clusterObjectsCheck
 	GetExpectedTemplateRefs(hostAwait *wait.HostAwaitility) TemplateRefs
+	GetNamespaceObjectChecks(nsType string) []namespaceObjectsCheck
 	GetTierObjectChecks() []tierObjectCheck
 }
 
+type customTierChecks struct {
+	getClusterObjectChecks   func() []clusterObjectsCheck
+	getExpectedTemplateRefs  func(hostAwait *wait.HostAwaitility) TemplateRefs
+	getNamespaceObjectChecks func(nsType string) []namespaceObjectsCheck
+	getTierObjectChecks      func() []tierObjectCheck
+}
+
+var _ TierChecks = &customTierChecks{}
+
+func newCustomChecks(tier *toolchainv1alpha1.NSTemplateTier) (TierChecks, error) {
+	// look-up annotations on the tier to figure out which "base" checks to use
+	c := &customTierChecks{}
+	baseTier, found := tier.Annotations[BaseTierKey]
+	if !found {
+		return nil, fmt.Errorf("no base assertion implementation found for %s", tier.Name)
+	}
+	baseChecks, err := newChecks(baseTier)
+	if err != nil {
+		return nil, err
+	}
+	c.getClusterObjectChecks = baseChecks.GetClusterObjectChecks
+	c.getExpectedTemplateRefs = baseChecks.GetExpectedTemplateRefs
+	c.getNamespaceObjectChecks = baseChecks.GetNamespaceObjectChecks
+	c.getTierObjectChecks = baseChecks.GetTierObjectChecks
+
+	// deal with replaced templaterefs for namespace and cluster resources
+	if tierName, found := tier.Annotations[ClusterResourcesBaseTierKey]; found {
+		checks, err := newChecks(tierName)
+		if err != nil {
+			return nil, err
+		}
+		c.getClusterObjectChecks = checks.GetClusterObjectChecks
+	}
+	if tierName, found := tier.Annotations[NamespaceResourcesBaseTierKey]; found {
+		checks, err := newChecks(tierName)
+		if err != nil {
+			return nil, err
+		}
+		c.getNamespaceObjectChecks = checks.GetNamespaceObjectChecks
+	}
+	return c, nil
+}
```

Comment:
using the annotation for storing the information on what tier was used for the creation of the new custom tier is additional hidden complexity that is not needed at all.
I mean - why do we need that in the first place? only to remove a few parameters from `verifyResourceUpdatesForUserSignups` function? If so, then I don't think that it's really worth it. Just compare the complexity/amount of code between this additional logic and the additional parameters in `verifyResourceUpdatesForUserSignups`.
Or is there anything else that I'm missing?

If you want to somehow somewhere store the information on what was used and create the checks based on that, then you could create a new type - let's say `CustomTier`:
```go
type CustomTier struct {
	tier                       *toolchainv1alpha1.NSTemplateTier
	namespacesSourceTier       *toolchainv1alpha1.NSTemplateTier
	clusterResourcesSourceTier *toolchainv1alpha1.NSTemplateTier
}
```
that would be returned from `CreateNSTemplateTier` function
The same could be then used in `updateNSTemplateTier` so it would update it with the actual tier that was used as the source of the templates.

and the same could be then passed to `verifyResourceUpdatesForUserSignups` for creating the needed checks. The main benefit compared to the annotation is that this is fully transparent.

Reply:
> I mean - why do we need that in the first place? only to remove a few parameters from verifyResourceUpdatesForUserSignups function? 

yes, because IMO the code is not easy to understand. It causes overhead/friction if one has to figure out what those parameters are and what they are used for. The goal of the PR was to remove this complexity. And since we've recently decided to opt for annotations in [some](https://github.com/codeready-toolchain/host-operator/pull/607) [specific](https://github.com/codeready-toolchain/host-operator/pull/595) [cases](https://github.com/codeready-toolchain/host-operator/pull/596), I thought we could also use this mechanism here, so we just need to pass the `tier` as an argument and let the `newCustomChecks` func figure out what actual checks are needed.

Reply:
There is one big difference between the cases your referenced and this one. The annotations are used to alter the behavior of the controller and are used in production. E2e tests then verify the expected behavior based on the given annotation.

However, in your case, you add annotations that are used purely in e2e tests. They don't alter anything in the controller behavior, they "only" mark what tiers were used when for the creation or for the update. As I mentioned in the example above - this can be easily solved by introducing a new type(s).
Using the new type would also solve the problem with the number of parameters - you would only pass one parameter - the `CustomTier` to be verified. 
I haven't tried to implement it, but I can imagine that in some cases you could have problem with re-using the same function for the cases when you need to pass `CustomTier` or `NSTemplateTier` - in such cases you can think of other concepts like taking not the `NSTemplateTier`, but directly the `TierChecks`...

What I'm trying to say - if the code is confusing, then let's try to make it simpler either via a new types or via fluent API or variadic function, but in a way that it's clear and transparent what it actually does. 
I have a feeling that with the annotation approach we just keep the code more or less the same (with less parameters), but hiding the complexity at places that are not transparent nor clear.

Reply:
I'm indeed looking at your `CustomTier` suggestion and indeed, it causes troubles with other funcs which expect an `NSTemplateTier` parameter. 
Once again, we disagree on what is clear and what is not.

Reply:
I understand that. 
So it's not only between you and me, let's ask the rest of the team what they would prefer:
* using annotation
* building either fluent API or custom types

@ranakan19 @rajivnathan @sbryzak @alexeykazakov 
But please, read the discussion and ideally also check the PR so you know the context.

Reply:
btw, another option that just came to my mind would be replacing the three parameters with one that would be called for example `TierContext` which could be created via fluent API:
```go
tierCtx := ForTier("cheesecake").NamespaceTemplatesBasedOn("advanced").ClusterResourcesBasedOn("base")
```
which would be used in:
```go
cheesecakeSyncIndexes = verifyResourceUpdatesForUserSignups(t, awaitilities, memberAwait, cheesecakeSyncIndexes, tierContext, false)
```

This shouldn't require so many changes and should be pretty straightforward. 
I just spent 15 minutes trying to quickly implement it to check if there is some hidden complication and it seems to be working fine (from the implementation point of view):
https://github.com/codeready-toolchain/toolchain-e2e/compare/master...MatousJobanek:tiercontext?expand=1

Reply:
I like the `TierContext` option!
I would try to minimize the refactoring as much as possible.

Reply:
@MatousJobanek you suggested al alternative solution with `CustomTier` so please let me take care of that now. And I prefer variadics to fluent api, but that's just my taste

Reply:
Based on the [previous comments](https://github.com/codeready-toolchain/toolchain-e2e/pull/486#discussion_r823779438) I got a feeling that you didn't want to spend the time with the `CustomTier` as it causes troubles with other funcs, so I proposed a much simpler solution that doesn't require so many changes and should solve the problem with the number of variables.
So I'm confused now... :man_shrugging: 

> I prefer variadics to fluent api, but that's just my taste

sure, no problem. Variadics is simpler from the implementation point of view. Fluent API is better as you know what functions/options are available and could be used. I'm fine with either way.

Reply:
I agree Matous. I'm still voting for the TierContext option.

Reply:
awesome.
how is the `TierContext` proposal better than the `CustomTier` one that Matous suggested above?
Also, `TierContext` is not a good name IMO.

Reply:
I missed the last comment, so sorry for the late response.
I thought that the `TierContext` would be better mainly because it was much simpler and it still solved the same problem. Since it would be simpler, then it would mean:
* fewer code changes
* shorter and quicker PR reviews
* faster feedback loop

which means that we would all spend less time with the refactoring effort. 
But I see that you already changed it in another way, so let me try to find some time for a review, but since it's bigger, then it will take a bit more time.

Reply:
I believe that the `CustomNSTemplateTier` type in my PR is quite close to the `TierContext` you suggested
There are other changes related to moving some funcs from `testsupport` to `testsupport/tiers` pkg, so it's only a matter of imports.

Reply:
Sorry for being late, I am catching up on PR reviews. Are we still deciding what to do? I guess Matous' `TierContext` idea would have been more focused on just combining the checks? Perhaps there's a better name but I do see how it can be simpler because we wouldn't have both `CustomNSTemplateTier` and `NSTemplateTier` to deal with it would just be about checks? That said, Xavier has already done the work for CustomTier and it's working now so 🤷‍♂️  If the work wasn't already done, I would have voted for the `TierContext` idea but now that it's done I'm not sure it'd be worth spending more time on this stuff.

Reply:
ok, let me see if there's a way to simply/factorize the way these `TierChecks` are initialized...

Reply:
@MatousJobanek @rajivnathan I refactored the checks in 23cefec. What I like about this solution is that we can more easily compare the checks between tiers, and the setup of the checks for the "CustomNSTemplateTier" is more straightforward. 
Please let me know what you think about it.